### PR TITLE
Merge Action Executions

### PR DESCRIPTION
**Pull request by werkt**
Multiple executions requested for the same action_digest may be merged. The global configuration can disable merging by default. If specified, an execute request's RequestMetadata may indicate with url-correlatedInvocationsId that it is enabled/disabled to defy default. After merging fails multiple times, the action is scheduled as a regular execution.
Merges exist as keys in backplane distributed state, and are only removed by workers or after an execution completion is 60s old (observed by server).

### CODE REVIEW

File: src/main/java/build/buildfarm/common/DigestUtil.java [hunk 1]
```diff
@@ -157,6 +157,11 @@ public boolean equals(Object o) {
     public int hashCode() {
       return digest.hashCode();
     }
+
+    @Override
+    public String toString() {
+      return DigestUtil.toString(digest);
+    }
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/common/DigestUtil.java:164]: Thank you!!!

File: src/main/java/build/buildfarm/instance/server/NodeInstance.java [hunk 2]
```diff
@@ -347,22 +347,35 @@ private static boolean shouldEnsureOutputsPresent(
...
+      QueryStringDecoder decoder = new QueryStringDecoder(uri);
+      return decoder
+          .parameters()
+          .getOrDefault(
+              "ENSURE_OUTPUTS_PRESENT", ImmutableList.of(ensureOutputsPresent ? "true" : "false"))
+          .getFirst()
+          .equals("true");
+    } catch (URISyntaxException e) {
+      return ensureOutputsPresent;
     }
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/instance/server/NodeInstance.java:364]: Could this be extracted to a function? Looks a lot like the next block in `shouldMergeExecutions(...)`

Comment (werkt) [reply src/main/java/build/buildfarm/instance/server/NodeInstance.java:364]: It also looks like the block in https://github.com/buildfarm/buildfarm/blob/main/src/main/java/build/buildfarm/server/InvocationsCollector.java#L82 for invocations. I probably just need to intercept this and put a map into the context to be extracted and sent to interested requests.

File: src/main/java/build/buildfarm/instance/shard/Executions.java [hunk 3]
```diff
@@ -63,15 +63,23 @@ static JsonFormat.Parser getParser() {
...
+  public RedisMap actions;
+
   /**
    * @brief Constructor.
    * @details Construct container for executions.
    * @param name The global name of the execution map.
    * @param timeout_s When to expire executions.
    */
-  public Executions(RedisSetMap toolInvocations, String name, int timeout_s) {
+  public Executions(RedisSetMap toolInvocations, String name, String actionsName, int timeout_s) {
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/instance/shard/Executions.java:79]: Add to javadoc

File: src/main/java/build/buildfarm/backplane/Backplane.java [hunk 4]
```diff
@@ -271,7 +271,12 @@ ScanResult<DispatchedOperation> scanDispatchedOperations(String cursor, int coun
   /** Requeue a dispatched execution */
   void requeueDispatchedExecution(QueueEntry queueEntry) throws IOException;
 
-  void prequeue(ExecuteEntry executeEntry, Operation operation) throws IOException;
+  Operation mergeExecution(ActionKey actionKey) throws IOException;
+
+  void unmergeExecution(ActionKey actionKey) throws IOException;
+
+  boolean prequeue(ExecuteEntry executeEntry, Operation operation, boolean ignoreMerge)
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/backplane/Backplane.java:278]: Javadoc? (at least the return value?)

File: src/main/java/build/buildfarm/instance/shard/Executions.java [hunk 5]
```diff
@@ -150,4 +167,16 @@ public void insert(UnifiedJedis jedis, String name, String operationJson) {
   public void remove(UnifiedJedis jedis, String name) {
     executions.remove(jedis, name);
   }
+
+  public Operation merge(UnifiedJedis jedis, String actionKey) {
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/instance/shard/Executions.java:171]:
```suggestion
  public @Nullable Operation merge(UnifiedJedis jedis, String actionKey) {
```
and import `@Nullable`

File: src/main/java/build/buildfarm/instance/shard/ServerInstance.java [hunk 6]
```diff
@@ -191,6 +192,8 @@ public class ServerInstance extends NodeInstance {
   // Prometheus metrics
   private static final Counter executionSuccess =
       Counter.build().name("execution_success").help("Execution success.").register();
+  private static final Counter mergedExecutions =
+      Counter.build().name("merged_executions").help("Merged executions.").register();
```
Comment (jasonschroeder-sfdc) [inline src/main/java/build/buildfarm/instance/shard/ServerInstance.java:196]: Document it in `_site/docs/metrics/metrics.md`

**Review by jasonschroeder-sfdc — APPROVED** [review]

**Pull request merged**